plugins/ports-plugin/src/devfile-handler-che-server-impl.ts

```diff
@@ -0,0 +1,198 @@
+/*********************************************************************
+ * Copyright (c) 2020 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ **********************************************************************/
+
+import { Endpoint, EndpointExposure, EndpointCategory } from './endpoint';
+import * as che from '@eclipse-che/plugin';
+import { CommandPreviewUrl } from './devfile-handler';
+
+/**
+ * Grab endpoints from a devfile.
+ * @author Florent Benoit
+ */
+
+export class CheServerDevfileHandlerImpl {
+    async getEndpoints(): Promise<Array<Endpoint>> {
+
+        const workspace = await che.workspace.getCurrentWorkspace();
+
+        const endpoints: Endpoint[] = [];
+        if (!workspace) {
+            return endpoints;
+        }
+
+        const previewUrls: CommandPreviewUrl[] = [];
+        if (workspace.devfile && workspace.devfile.commands) {
+            const commands = workspace.devfile.commands;
+            for (const command of commands) {
+                if (command.previewUrl && command.previewUrl.port) {
+                    previewUrls.push({ port: command.previewUrl.port, path: command.previewUrl.path });
+                }
+            }
+        }
+
+        const runtimeMachines = workspace.runtime!.machines || {};
```

Comment:
how could it be undefined ?

Reply:
Because the type says it can be undefined and we didn't check it. If it can't be undefined, we should fix the types.

Reply:
I mean, type is correct, because I can ask information about other workspaces.
But here we ask the runtime of the current workspace, and if runtime attribute is not there it means that workspace is stopped. And if it's stopped it means that theia will not be started

Reply:
If we need that much context knowledge to understand the code, it would warrant a comment.